lib/puppet/provider/service/smf.rb

```diff
@@ -2,117 +2,139 @@
 
 # Solaris 10 SMF-style services.
 Puppet::Type.type(:service).provide :smf, :parent => :base do
+  # TODO: Update the documentation with things that you find.
+  # E.g. About specifying fmri explicitly in the manifest.
   desc <<-EOT
     Support for Sun's new Service Management Framework.
 
-    Starting a service is effectively equivalent to enabling it, so there is
-    only support for starting and stopping services, which also enables and
-    disables them, respectively.
-
     By specifying `manifest => "/path/to/service.xml"`, the SMF manifest will
     be imported if it does not exist.
-
   EOT
 
   defaultfor :osfamily => :solaris
 
   confine :osfamily => :solaris
 
-  commands :adm => "/usr/sbin/svcadm", :svcs => "/usr/bin/svcs"
-  commands :svccfg => "/usr/sbin/svccfg"
+  commands :adm => "/usr/sbin/svcadm",
+           :svcs => "/usr/bin/svcs",
+           :svccfg => "/usr/sbin/svccfg"
 
   has_feature :refreshable
 
-  def setupservice
-      if resource[:manifest]
-        begin
-          svcs("-l", @resource[:name])
-        rescue Puppet::ExecutionFailure
-          Puppet.notice "Importing #{@resource[:manifest]} for #{@resource[:name]}"
-          svccfg :import, resource[:manifest]
-        end
-      end
+  def self.instances
+   service_instances = svcs("-H", "-o", "state,fmri" ).split("\n")
+
+   # Puppet does not manage services in the legacy_run state, so filter those out.
+   service_instances.reject! { |line| line =~ /^legacy_run/ }
+   
+   service_instances.collect! do |line|
+     _state, fmri = line.split(/\s+/)
+     new({:name => fmri})
+   end
+
+   service_instances
+  end
+
+  def initialize(*args)
+    super(*args)
+
+    # This hash contains the properties we need to sync. in our flush method. We use a separate hash
+    # instead of @properties_hash to simplify its implementation, since the latter stores ALL of the
+    # specified properties, not just the ones that need to be syncd.
+    @properties_to_sync = {}
```

Comment:
Does this seem reasonable? `properties_hash` seemed a little overkill and it seems like not using it would be not be harmful.

Reply:
That's a typo on my part, should be `@property_hash`. I'll update after the initial review. I don't see anything that relies on `@property_hash`, b/c from what I can tell `@property_hash` is correctly updated when the params. get synced and we sync our params. in the `start/stop`, `enable/disable` methods so that part is fine (+ Puppet correctly prints out that the params. have changed).

The other uses of `@property_hash` in the `#flush` method (that's most closely similar to this) correspond to checking the current values on top of the final values. See e.g. https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/provider/vlan/cisco.rb#L24 and https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/provider/network_device.rb#L65-L71 is where `former_properties` and `properties` are computed. `former_properties` is just the `property_hash` before syncing, `properties` is the `property_hash` after syncing. We don't really care about the old properties here, just what's changed so that's why I was thinking it's safe to store this in a separate variable.

Reply:
Coolness. That makes better sense.